tests/ssg_test_suite/rule.py

```diff
@@ -344,16 +363,45 @@ def _check_and_record_rule_scenario(self, scenario, remote_rule_dir, rule_id, re
         self._current_result.scenario = common.Scenario_run(rule_id, scenario.script)
         self._current_result.when = self.test_timestamp_str
 
-        self._check_rule_scenario(scenario, remote_rule_dir, rule_id, remediation_available)
+        with self.copy_of_datastream():
+            self._check_rule_scenario(scenario, remote_rule_dir, rule_id, remediation_available)
         self.results.append(self._current_result.save_to_dict())
 
+    @contextlib.contextmanager
+    def copy_of_datastream(self, new_filename=None):
+        old_filename = self.datastream
+        if not new_filename:
+            _, new_filename = tempfile.mkstemp(prefix="ssgts_ds_modified", dir="/tmp")
```

Comment:
Why is a temporary datastream created once here and second time in _change_variable_value? How does it work?

Reply:
Copying of datastereams works as follows:

- Copy upon start of the test suite: When the test suite starts, it copies the datastream, so the user can e.g. rebuild without affecting tests.
- Copy before executing a rule scenario: Newly introduced, this code assumes that any test scenario may make edits to the datastream that shouldn't be persistent, so every scenario just gets a copy.
- Copy during value change: That one is just an implementation detail - it wouldn't be needed if `xsltproc` could do an in-place modification.